File: sdk/typescript/src/workflow-define.ts
/**
 * Fluent workflow definition builder with proxy-based reference capture.
 *
 * This is a structural fluent builder: references are captured into workflow
 * value nodes, but step IDs and paths are not accumulated as compile-time types.
 *
 * @module
 */

import type { AgentOutput, AgentToolRef } from "./providers/agent.ts";
import type { JsonInput, JsonObjectInput } from "./protocol.ts";
import {
  boundWorkflowTarget,
  workflowActivation,
  workflowAgentMessage,
  workflowDefinitionSpec,
  workflowEventActivation,
  workflowEventMatch,
  workflowStep,
  workflowStepAgentTurn,
  workflowStepAppCall,
  workflowStepWhen,
  workflowText,
  workflowValue,
  type BoundWorkflowTarget,
  type WorkflowActivation,
  type WorkflowDefinitionSpec,
  type WorkflowStep,
  type WorkflowValue,
} from "./providers/workflow.ts";

const WORKFLOW_REF = Symbol.for("gestalt.workflow.ref");

type WorkflowRefKind = "input" | "signal" | "stepOutput" | "stepInput";

interface WorkflowRef {
  readonly [WORKFLOW_REF]: true;
  readonly kind: WorkflowRefKind;
  readonly path: string;
  readonly stepId?: string | undefined;
}

interface WorkflowLiteralRef {
  readonly [WORKFLOW_REF]: true;
  readonly kind: "literal";
  readonly value: JsonInput;
}

interface WorkflowTemplateRef {
  readonly [WORKFLOW_REF]: true;
  readonly kind: "template";
  readonly template: string;
}

interface WorkflowTextExpression {
  readonly [WORKFLOW_REF]: true;
  readonly kind: "composedText";
  readonly parts: ReadonlyArray<string | WorkflowRef>;
}

type CapturedRef = WorkflowRef | WorkflowLiteralRef | WorkflowTemplateRef | WorkflowTextExpression;

export type WorkflowDefinitionInput =
  | WorkflowDefinitionSpec
  | WorkflowBuilder
  | { toSpec(): WorkflowDefinitionSpec };

export type WorkflowPrompt =
  | string
  | WorkflowTextExpression
  | ((scope: StepScope) => string | WorkflowTextExpression);

export interface DefineWorkflowOptions {
  id: string;
  runAs: string;
  paused?: boolean | undefined;
}

export interface EventActivationOptions {
  id?: string | undefined;
  source?: string | undefined;
  subject?: string | undefined;
  paused?: boolean | undefined;
}

export interface ScheduleActivationOptions {
  id?: string | undefined;
  timezone?: string | undefined;
  paused?: boolean | undefined;
}

export interface EventActivationConfig {
  readonly __workflowActivation: "event";
  readonly type: string;
  readonly mapInput?: ((event: EventScope) => Record<string, unknown>) | undefined;
  readonly options?: EventActivationOptions | undefined;
}

export interface ScheduleActivationConfig {
  readonly __workflowActivation: "schedule";
  readonly cron: string;
  readonly mapInput?: ((scope: ActivationScope) => Record<string, unknown>) | undefined;
  readonly options?: ScheduleActivationOptions | undefined;
}

export type ActivationConfig = EventActivationConfig | ScheduleActivationConfig;

export interface StepAppConfig {
  name: string;
  operation: string;
  input?: ((scope: StepScope) => Record<string, unknown>) | undefined;
  connection?: string | undefined;
  instance?: string | undefined;
  credentialMode?: string | undefined;
}

export interface StepAgentMessageConfig {
  role: string;
  text: WorkflowPrompt;
}

export interface StepAgentConfig {
  provider: string;
  model?: string | undefined;
  sessionKey?: string | undefined;
  prompt?: WorkflowPrompt | undefined;
  messages?: readonly StepAgentMessageConfig[] | undefined;
  tools?: readonly AgentToolRef[] | undefined;
  output?: AgentOutput | undefined;
  modelOptions?: JsonObjectInput | undefined;
}

export interface StepWhenConfig {
  value: unknown;
  equals?: JsonInput | undefined;
}

export interface StepConfig {
  app?: StepAppConfig | undefined;
  agent?: StepAgentConfig | undefined;
  inputs?: Record<string, unknown> | ((scope: StepScope) => Record<string, unknown>) | undefined;
  when?: StepWhenConfig | undefined;
  timeoutSeconds?: number | undefined;
  metadata?: JsonObjectInput | undefined;
}

export interface StepRefProxy {
  readonly [key: string]: unknown;
}

export interface StepScopeRefs {
  readonly output: StepRefProxy;
  readonly input: StepRefProxy;
}

export interface StepScope {
  readonly input: StepRefProxy;
  readonly signal: StepRefProxy;
  stepOutput(stepId: string, path: string): WorkflowRef;
  stepInput(stepId: string, path: string): WorkflowRef;
  readonly steps: Record<string, StepScopeRefs>;
}

export type EventScope = StepRefProxy;
export type ActivationScope = StepRefProxy;

/** Creates an event activation configuration for `.on()`. */
export function event(
  type: string,
  mapInput?: (event: EventScope) => Record<string, unknown>,
  options?: EventActivationOptions,
): EventActivationConfig {
  return {
    __workflowActivation: "event",
    type,
    mapInput,
    options,
  };
}

/** Creates a schedule activation configuration for `.on()`. */
export function schedule(
  cron: string,
  mapInput?: (scope: ActivationScope) => Record<string, unknown>,
  options?: ScheduleActivationOptions,
): ScheduleActivationConfig {
  return {
    __workflowActivation: "schedule",
    cron,
    mapInput,
    options,
  };
}

/** Starts a fluent workflow definition builder. `runAs` is required. */
export function defineWorkflow(options: DefineWorkflowOptions): WorkflowBuilder {
  const runAs = options.runAs?.trim() ?? "";
  if (!runAs) {
    throw new Error("defineWorkflow requires runAs");
  }
  const id = options.id?.trim() ?? "";
  if (!id) {
    throw new Error("defineWorkflow requires id");
  }
  return new WorkflowBuilder({
    id,
    runAs,
    paused: options.paused ?? false,
    activations: [],
    steps: [],
  });
}

/** Composes workflow prompt/message text from literals and captured references. */
export function text(...parts: Array<string | StepRefProxy | WorkflowRef>): WorkflowTextExpression {
  const normalized: Array<string | WorkflowRef> = [];
  for (const part of parts) {
    if (typeof part === "string") {
      normalized.push(part);
      continue;
    }
    if (isWorkflowRef(part)) {
      normalized.push(part);
      continue;
    }
    throw new Error("text() parts must be strings or workflow references");
  }
  return {
    [WORKFLOW_REF]: true,
    kind: "composedText",
    parts: normalized,
  };
}

/** Returns a workflow definition spec when a builder is passed where a raw spec is accepted. */
export function resolveWorkflowDefinitionSpec(input: WorkflowDefinitionInput): WorkflowDefinitionSpec {
  if (input instanceof WorkflowBuilder) {
    return input.toSpec();
  }
  if (
    typeof input === "object" &&
    input !== null &&
    "toSpec" in input &&
    typeof input.toSpec === "function"
  ) {
    return input.toSpec();
  }
  return workflowDefinitionSpec(input as WorkflowDefinitionSpec);
}

export class WorkflowBuilder {
  private readonly state: {
    id: string;
    runAs: string;
    paused: boolean;
    activations: WorkflowActivation[];
    steps: WorkflowStep[];
  };

  constructor(state: WorkflowBuilder["state"]) {
    this.state = state;
  }

  on(activation: ActivationConfig): this {
    if (activation.__workflowActivation === "event") {
      const activationId = activation.options?.id?.trim() || activation.type;
      const mapped = activation.mapInput
        ? activation.mapInput(createRefProxy("", "signal") as EventScope)
        : undefined;
      this.state.activations.push(
        workflowActivation({
          id: activationId,
          paused: activation.options?.paused ?? false,
          event: workflowEventActivation({
            match: workflowEventMatch({
              type: activation.type,
              source: activation.options?.source ?? "",
              subject: activation.options?.subject ?? "",
            }),
          }),
          input:
            mapped === undefined
              ? undefined
              : workflowValue({ object: captureObject(mapped, "signal") }),
        }),
      );
      return this;
    }

    const activationId = activation.options?.id?.trim() || activation.cron;
    const mapped = activation.mapInput
      ? activation.mapInput(createRefProxy("", "input") as ActivationScope)
      : undefined;
    this.state.activations.push(
      workflowActivation({
        id: activationId,
        paused: activation.options?.paused ?? false,
        schedule: {
          cron: activation.cron,
          timezone: activation.options?.timezone ?? "",
        },
        input:
          mapped === undefined
            ? undefined
            : workflowValue({ object: captureObject(mapped, "input") }),
      }),
    );
    return this;
  }

  step(stepId: string, config: StepConfig): this {
    const scope = createStepScope();
    const step: WorkflowStep = { id: stepId };

    if (config.inputs !== undefined) {
      const mapped =
        typeof config.inputs === "function" ? config.inputs(scope) : config.inputs;
      step.inputs = captureObject(mapped, "input");
    }

    if (config.app !== undefined) {
      const input = config.app.input?.(scope);
      step.app = workflowStepAppCall({
        name: config.app.name,
        operation: config.app.operation,
        input:
          input === undefined ? undefined : workflowValue({ object: captureObject(input, "input") }),
        connection: config.app.connection,
        instance: config.app.instance,
        credentialMode: config.app.credentialMode,
      });
    }

    if (config.agent !== undefined) {
      const prompt =
        typeof config.agent.prompt === "function"
          ? config.agent.prompt(scope)
          : config.agent.prompt;
      step.agent = workflowStepAgentTurn({
        provider: config.agent.provider,
        model: config.agent.model,
        sessionKey: config.agent.sessionKey,
        prompt:
          prompt === undefined ? undefined : workflowText(resolveWorkflowPrompt(prompt)),
        messages: config.agent.messages?.map((message) => {
          const textValue =
            typeof message.text === "function" ? message.text(scope) : message.text;
          return workflowAgentMessage({
            role: message.role,
            text: workflowText(resolveWorkflowPrompt(textValue)),
          });
        }),
        tools: config.agent.tools ? [...config.agent.tools] : undefined,
        output: config.agent.output,
        modelOptions: config.agent.modelOptions,
      });
    }

    if (config.when !== undefined) {
      const whenValue =
        typeof config.when.value === "object" &&
        config.when.value !== null &&
        "kind" in (config.when.value as WorkflowValue)
          ? workflowValue(config.when.value as WorkflowValue)
          : workflowValue(captureValue(config.when.value, "input"));
      step.when = workflowStepWhen({
        value: whenValue,
        equals: config.when.equals,
      });
    }

    if (config.timeoutSeconds !== undefined) {
      step.timeoutSeconds = config.timeoutSeconds;
    }
    if (config.metadata !== undefined) {
      step.metadata = config.metadata;
    }

    this.state.steps.push(workflowStep(step));
    return this;
  }

  toSpec(): WorkflowDefinitionSpec {
    const target: BoundWorkflowTarget | undefined =
      this.state.steps.length === 0
        ? undefined
        : boundWorkflowTarget({ steps: this.state.steps });
    return workflowDefinitionSpec({
      id: this.state.id,
      runAs: this.state.runAs,
      paused: this.state.paused,
      activations: this.state.activations,
      target,
    });
  }
}

function createStepScope(): StepScope {
  return {
    input: createRefProxy("", "input") as StepRefProxy,
    signal: createRefProxy("", "signal") as StepRefProxy,
    stepOutput(stepId: string, path: string): WorkflowRef {
      return {
        [WORKFLOW_REF]: true,
        kind: "stepOutput",
        stepId,
        path,
      };
    },
    stepInput(stepId: string, path: string): WorkflowRef {
      return {
        [WORKFLOW_REF]: true,
        kind: "stepInput",
        stepId,
        path,
      };
    },
    steps: new Proxy(
      {},
      {
        get(_target, stepId) {
          if (typeof stepId !== "string") {
            return undefined;
          }
          return {
            output: createRefProxy("", "stepOutput", stepId) as StepRefProxy,
            input: createRefProxy("", "stepInput", stepId) as StepRefProxy,
          };
        },
      },
    ),
  };
}

function createRefProxy(
  path: string,
  kind: WorkflowRefKind,
  stepId?: string,
): StepRefProxy {
  const marker: WorkflowRef = {
    [WORKFLOW_REF]: true,
    kind,
    path,
    stepId,
  };
  return new Proxy(marker, {
    get(target, prop, receiver) {
      if (prop === WORKFLOW_REF) {
        return true;
      }
      if (prop === "kind") {
        return target.kind;
      }
      if (prop === "path") {
        return target.path;
      }
      if (prop === "stepId") {
        return target.stepId;
      }
      if (typeof prop === "symbol") {
        return Reflect.get(target, prop, receiver);
      }
      const nextPath = target.path ? `${target.path}.${String(prop)}` : String(prop);
      return createRefProxy(nextPath, target.kind, target.stepId);
    },
  }) as unknown as StepRefProxy;
}

function isCapturedRef(value: unknown): value is CapturedRef {
  return (
    typeof value === "object" &&
    value !== null &&
    (value as CapturedRef)[WORKFLOW_REF] === true
  );
}

function isWorkflowRef(value: unknown): value is WorkflowRef {
  return (
    isCapturedRef(value) &&
    (value.kind === "input" ||
      value.kind === "signal" ||
      value.kind === "stepOutput" ||
      value.kind === "stepInput")
  );
}

function isTextExpression(value: unknown): value is WorkflowTextExpression {
  return isCapturedRef(value) && value.kind === "composedText";
}

function resolveWorkflowPrompt(value: string | WorkflowTextExpression): string {
  if (typeof value === "string") {
    return value;
  }
  if (isTextExpression(value)) {
    return renderTextExpression(value);
  }
  throw new Error("workflow prompt must be a string or text() expression");
}

function renderTextExpression(expression: WorkflowTextExpression): string {
  let rendered = "";
  for (const part of expression.parts) {
    if (typeof part === "string") {
      rendered += part;
      continue;
    }
    rendered += refToTemplatePlaceholder(part);
  }
  return rendered;
}

function refToTemplatePlaceholder(ref: WorkflowRef): string {
  switch (ref.kind) {
    case "input":
      return `\${{ input.${ref.path} }}`;
    case "signal":
      return `\${{ signal.${ref.path} }}`;
    case "stepOutput":
      return `\${{ steps.${ref.stepId ?? ""}.outputs.${ref.path} }}`;
    case "stepInput":
      return `\${{ steps.${ref.stepId ?? ""}.inputs.${ref.path} }}`;
  }
}

function captureObject(
  value: Record<string, unknown>,
  defaultKind: WorkflowRefKind,
): Record<string, WorkflowValue> {
  const out: Record<string, WorkflowValue> = {};
  for (const [key, nested] of Object.entries(value)) {
    out[key] = workflowValue(captureValue(nested, defaultKind));
  }
  return out;
}

function captureValue(value: unknown, defaultKind: WorkflowRefKind): WorkflowValue {
  if (isCapturedRef(value)) {
    if (value.kind === "composedText") {
      throw new Error("workflow values cannot use text(); use text() only for prompts and messages");
    }
    switch (value.kind) {
      case "input":
        return { input: value.path };
      case "signal":
        return { signal: value.path };
      case "stepOutput":
        return { stepOutput: { stepId: value.stepId ?? "", path: value.path } };
      case "stepInput":
        return { stepInput: { stepId: value.stepId ?? "", path: value.path } };
      case "literal":
        return { literal: value.value };
      case "template":
        return { template: workflowText(value.template) };
      default: {
        const exhaustive: never = value;
        throw new Error(`unsupported workflow ref kind: ${String(exhaustive)}`);
      }
    }
  }

  if (Array.isArray(value)) {
    return {
      array: value.map((item) => workflowValue(captureValue(item, defaultKind))),
    };
  }

  if (value !== null && typeof value === "object") {
    return {
      object: captureObject(value as Record<string, unknown>, defaultKind),
    };
  }

  return { literal: value as JsonInput };
}

File: sdk/python/gestalt/workflow_define.py
"""Fluent workflow definition builder with proxy-based reference capture."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Mapping, Protocol

from gestalt._gen.v1 import workflow_pb2 as pb
from gestalt._workflow import (
    bound_workflow_target,
    workflow_activation,
    workflow_agent_message,
    workflow_definition_spec,
    workflow_event_match,
    workflow_step,
    workflow_step_agent_turn,
    workflow_step_app_call,
    workflow_text,
    workflow_value,
)


class _TextExpression:
    __slots__ = ("parts",)

    def __init__(self, parts: tuple[str | _WorkflowRef | _WorkflowRefProxy, ...]) -> None:
        self.parts = parts


class _TemplateMarker:
    __slots__ = ("template",)

    def __init__(self, template: str) -> None:
        self.template = template


class _WorkflowRef:
    __slots__ = ("kind", "path", "step_id")

    def __init__(self, kind: str, path: str = "", step_id: str = "") -> None:
        self.kind = kind
        self.path = path
        self.step_id = step_id


class _WorkflowRefProxy:
    def __init__(self, kind: str, path: str = "", step_id: str = "") -> None:
        self._ref = _WorkflowRef(kind, path, step_id)

    def __getattr__(self, name: str) -> _WorkflowRefProxy:
        next_path = f"{self._ref.path}.{name}" if self._ref.path else name
        return _WorkflowRefProxy(self._ref.kind, next_path, self._ref.step_id)

    def __getitem__(self, name: str) -> _WorkflowRefProxy:
        return self.__getattr__(str(name))


class _StepScope:
    def __init__(self) -> None:
        self.input = _WorkflowRefProxy("input")
        self.signal = _WorkflowRefProxy("signal")
        self._steps: dict[str, _StepOutputs] = {}

    def step_output(self, step_id: str, path: str) -> _WorkflowRef:
        return _WorkflowRef("stepOutput", path, step_id)

    def step_input(self, step_id: str, path: str) -> _WorkflowRef:
        return _WorkflowRef("stepInput", path, step_id)

    @property
    def steps(self) -> _StepOutputsMap:
        return _StepOutputsMap(self)


class _StepOutputs:
    def __init__(self, step_id: str) -> None:
        self.output = _WorkflowRefProxy("stepOutput", "", step_id)
        self.input = _WorkflowRefProxy("stepInput", "", step_id)


class _StepOutputsMap:
    def __init__(self, scope: _StepScope) -> None:
        self._scope = scope

    def __getitem__(self, step_id: str) -> _StepOutputs:
        return _StepOutputs(step_id)


class _EventScope:
    def __init__(self) -> None:
        self.data = _WorkflowRefProxy("signal", "data")


@dataclass(frozen=True)
class EventActivationConfig:
    type: str
    map_input: Callable[[_EventScope], Mapping[str, Any]] | None = None
    id: str = ""
    source: str = ""
    subject: str = ""
    paused: bool = False


@dataclass(frozen=True)
class ScheduleActivationConfig:
    cron: str
    map_input: Callable[[_WorkflowRefProxy], Mapping[str, Any]] | None = None
    id: str = ""
    timezone: str = ""
    paused: bool = False


class WorkflowBuilder:
    def __init__(self, *, workflow_id: str, run_as: str, paused: bool = False) -> None:
        if not run_as.strip():
            raise ValueError("define_workflow requires run_as")
        if not workflow_id.strip():
            raise ValueError("define_workflow requires id")
        self._id = workflow_id
        self._run_as = run_as
        self._paused = paused
        self._activations: list[Any] = []
        self._steps: list[Any] = []

    def on(self, activation: EventActivationConfig | ScheduleActivationConfig) -> WorkflowBuilder:
        if isinstance(activation, EventActivationConfig):
            activation_id = activation.id.strip() or activation.type
            mapped = (
                activation.map_input(_EventScope()) if activation.map_input is not None else None
            )
            self._activations.append(
                workflow_activation(
                    id=activation_id,
                    paused=activation.paused,
                    event={"match": workflow_event_match(
                        type=activation.type,
                        source=activation.source,
                        subject=activation.subject,
                    )},
                    input=None
                    if mapped is None
                    else _capture_object_value(mapped, "signal"),
                )
            )
            return self

        activation_id = activation.id.strip() or activation.cron
        mapped = (
            activation.map_input(_WorkflowRefProxy("input"))
            if activation.map_input is not None
            else None
        )
        self._activations.append(
            workflow_activation(
                id=activation_id,
                paused=activation.paused,
                schedule={"cron": activation.cron, "timezone": activation.timezone},
        input=None
        if mapped is None
        else _capture_object_value(mapped, "input"),
            )
        )
        return self

    def step(self, step_id: str, config: Mapping[str, Any]) -> WorkflowBuilder:
        scope = _StepScope()
        step_kwargs: dict[str, Any] = {"id": step_id}

        inputs = config.get("inputs")
        if inputs is not None:
            mapped = inputs(scope) if callable(inputs) else inputs
            step_kwargs["inputs"] = _capture_object(mapped, "input")

        app = config.get("app")
        if app is not None:
            app_input = app.get("input")
            mapped_input = app_input(scope) if callable(app_input) else app_input
            step_kwargs["app"] = workflow_step_app_call(
                name=app.get("name", ""),
                operation=app.get("operation", ""),
                input=None
                if mapped_input is None
                else _capture_object_value(mapped_input, "input"),
                connection=app.get("connection", ""),
                instance=app.get("instance", ""),
                credential_mode=app.get("credential_mode", ""),
            )

        agent = config.get("agent")
        if agent is not None:
            prompt = agent.get("prompt")
            if callable(prompt):
                prompt = prompt(scope)
            messages = []
            for message in agent.get("messages", []):
                message_text = message.get("text")
                if callable(message_text):
                    message_text = message_text(scope)
                messages.append(
                    workflow_agent_message(
                        role=message.get("role", ""),
                        text=workflow_text(_resolve_workflow_prompt(message_text)),
                    )
                )
            step_kwargs["agent"] = workflow_step_agent_turn(
                provider=agent.get("provider", ""),
                model=agent.get("model", ""),
                session_key=agent.get("session_key", ""),
                prompt=None if prompt is None else workflow_text(_resolve_workflow_prompt(prompt)),
                messages=messages,
                tools=agent.get("tools"),
                output=agent.get("output"),
                model_options=agent.get("model_options"),
            )

        when = config.get("when")
        if when is not None:
            value = when.get("value")
            proto_value = _capture_proto(value, "input")
            condition = pb.WorkflowStepWhen(value=proto_value)
            if "equals" in when:
                from gestalt._workflow import _value

                condition.equals.CopyFrom(_value(when.get("equals")))
            step_kwargs["when"] = condition

        if "timeout_seconds" in config:
            step_kwargs["timeout_seconds"] = config["timeout_seconds"]
        if "metadata" in config:
            step_kwargs["metadata"] = config["metadata"]

        self._steps.append(workflow_step(**step_kwargs))
        return self

    def to_spec(self) -> Any:
        target = None
        if self._steps:
            target = bound_workflow_target(steps=self._steps)
        return workflow_definition_spec(
            id=self._id,
            run_as=self._run_as,
            paused=self._paused,
            activations=self._activations,
            target=target,
        )


def define_workflow(*, workflow_id: str, run_as: str, paused: bool = False) -> WorkflowBuilder:
    return WorkflowBuilder(workflow_id=workflow_id, run_as=run_as, paused=paused)


def event(
    type_name: str,
    map_input: Callable[[_EventScope], Mapping[str, Any]] | None = None,
    *,
    id: str = "",
    source: str = "",
    subject: str = "",
    paused: bool = False,
) -> EventActivationConfig:
    return EventActivationConfig(
        type=type_name,
        map_input=map_input,
        id=id,
        source=source,
        subject=subject,
        paused=paused,
    )


def schedule(
    cron: str,
    map_input: Callable[[_WorkflowRefProxy], Mapping[str, Any]] | None = None,
    *,
    id: str = "",
    timezone: str = "",
    paused: bool = False,
) -> ScheduleActivationConfig:
    return ScheduleActivationConfig(
        cron=cron,
        map_input=map_input,
        id=id,
        timezone=timezone,
        paused=paused,
    )


def text(*parts: str | _WorkflowRef | _WorkflowRefProxy) -> _TextExpression:
    """Compose workflow prompt/message text from literals and captured references."""
    return _TextExpression(parts)


class _WorkflowDefinitionSpecProvider(Protocol):
    def to_spec(self) -> Any: ...


def resolve_workflow_definition_spec(spec: WorkflowBuilder | _WorkflowDefinitionSpecProvider | Mapping[str, Any] | Any) -> Any:
    if isinstance(spec, WorkflowBuilder):
        return spec.to_spec()
    to_spec = getattr(spec, "to_spec", None)
    if callable(to_spec):
        return workflow_definition_spec(to_spec())
    return workflow_definition_spec(spec)


def _capture_object(
    value: Mapping[str, Any], default_kind: str
) -> dict[str, Any]:
    return {
        key: _capture_proto(item, default_kind)
        for key, item in value.items()
    }


def _capture_object_value(value: Mapping[str, Any], default_kind: str) -> Any:
    return pb.WorkflowValue(
        object=pb.WorkflowObject(fields=_capture_object(value, default_kind))
    )


def _capture_proto(value: Any, default_kind: str) -> Any:
    if isinstance(value, _WorkflowRef):
        if value.kind == "stepOutput":
            return pb.WorkflowValue(
                step_output=pb.WorkflowStepOutputSource(
                    step_id=value.step_id, path=value.path
                )
            )
        if value.kind == "stepInput":
            return pb.WorkflowValue(
                step_input=pb.WorkflowStepInputSource(
                    step_id=value.step_id, path=value.path
                )
            )
        if value.kind == "input":
            return pb.WorkflowValue(input=pb.WorkflowPathSource(path=value.path))
        if value.kind == "signal":
            return pb.WorkflowValue(signal=pb.WorkflowPathSource(path=value.path))
    if isinstance(value, _WorkflowRefProxy):
        ref = value._ref
        if ref.kind == "stepOutput":
            return pb.WorkflowValue(
                step_output=pb.WorkflowStepOutputSource(
                    step_id=ref.step_id, path=ref.path
                )
            )
        if ref.kind == "stepInput":
            return pb.WorkflowValue(
                step_input=pb.WorkflowStepInputSource(
                    step_id=ref.step_id, path=ref.path
                )
            )
        if ref.kind == "input":
            return pb.WorkflowValue(input=pb.WorkflowPathSource(path=ref.path))
        if ref.kind == "signal":
            return pb.WorkflowValue(signal=pb.WorkflowPathSource(path=ref.path))
    if isinstance(value, _TemplateMarker):
        return workflow_value(template=value.template)
    if isinstance(value, list):
        return pb.WorkflowValue(
            array=pb.WorkflowArray(
                values=[_capture_proto(item, default_kind) for item in value]
            )
        )
    if isinstance(value, Mapping):
        return pb.WorkflowValue(
            object=pb.WorkflowObject(
                fields={
                    key: _capture_proto(item, default_kind)
                    for key, item in value.items()
                }
            )
        )
    return workflow_value(literal=value)


def _resolve_workflow_prompt(value: Any) -> str:
    if isinstance(value, str):
        return value
    if isinstance(value, _TextExpression):
        return _render_text_expression(value)
    raise ValueError("workflow prompt must be a string or text() expression")


def _render_text_expression(expression: _TextExpression) -> str:
    rendered = ""
    for part in expression.parts:
        if isinstance(part, str):
            rendered += part
            continue
        rendered += _ref_to_template_placeholder(part)
    return rendered


def _ref_to_template_placeholder(ref: _WorkflowRef | _WorkflowRefProxy) -> str:
    if isinstance(ref, _WorkflowRefProxy):
        ref = ref._ref
    if ref.kind == "input":
        return f"${{{{ input.{ref.path} }}}}"
    if ref.kind == "signal":
        return f"${{{{ signal.{ref.path} }}}}"
    if ref.kind == "stepOutput":
        return f"${{{{ steps.{ref.step_id}.outputs.{ref.path} }}}}"
    if ref.kind == "stepInput":
        return f"${{{{ steps.{ref.step_id}.inputs.{ref.path} }}}}"
    raise ValueError(f"unsupported workflow ref kind: {ref.kind}")
